adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleInterstitial.m

```diff
@@ -91,98 +93,74 @@ - (void)requestInterstitialAd {
 #pragma mark - GADMediationInterstitialAd Methods
 
 - (void)presentFromViewController:(UIViewController *)rootViewController {
-  NSError *error = nil;
-  if (![VungleSDK.sharedSDK
-               playAd:rootViewController
-              options:GADMAdapterVunglePlaybackOptionsDictionaryForExtras(_adConfiguration.extras)
-          placementID:self.desiredPlacement
-             adMarkup:[self bidResponse]
-                error:&error]) {
-    // Ad not playable.
-    if (error) {
-      [_delegate didFailToPresentWithError:error];
-    }
-  }
+  [_interstitialAd presentWith:rootViewController];
 }
 
 #pragma mark - Private methods
 
 - (void)loadAd {
-  NSError *error = [GADMAdapterVungleBiddingRouter.sharedInstance loadAdWithDelegate:self];
-  if (error) {
-    _adLoadCompletionHandler(nil, error);
-  }
+  _interstitialAd = [[VungleInterstitial alloc] initWithPlacementId:self.desiredPlacement];
+  _interstitialAd.delegate = self;
+  [_interstitialAd load:_adConfiguration.bidResponse];
 }
 
-#pragma mark - GADMAdapterVungleDelegate
-
-- (NSString *)bidResponse {
-  return _adConfiguration.bidResponse;
-}
-
-- (void)initialized:(BOOL)isSuccess error:(nullable NSError *)error {
-  if (!isSuccess) {
-    _adLoadCompletionHandler(nil, error);
-    return;
-  }
-  [self loadAd];
-}
-
-- (void)adAvailable {
-  if (self.isAdLoaded) {
-    // Already invoked an ad load callback.
-    return;
-  }
-  self.isAdLoaded = YES;
+#pragma mark - VungleInterstitialDelegate
 
+- (void)interstitialAdDidLoad:(VungleInterstitial *)interstitial {
   if (_adLoadCompletionHandler) {
     _delegate = _adLoadCompletionHandler(self, nil);
   }
-
-  if (!_delegate) {
-    // In this case, the request for Vungle has been timed out. Clean up self.
-    [GADMAdapterVungleBiddingRouter.sharedInstance removeDelegate:self];
-  }
 }
 
-- (void)adNotAvailable:(nonnull NSError *)error {
-  if (self.isAdLoaded) {
-    // Already invoked an ad load callback.
-    return;
-  }
-  _adLoadCompletionHandler(nil, error);
+- (void)interstitialAdDidFailToLoad:(VungleInterstitial *)interstitial withError:(NSError *)error {
+  NSError *gadError = GADMAdapterVungleErrorToGADError(GADMAdapterVungleErrorAdNotPlayable,
+                                                       error.code, error.localizedDescription);
+  _adLoadCompletionHandler(nil, gadError);
 }
 
-- (void)willShowAd {
+- (void)interstitialAdWillPresent:(VungleInterstitial *)interstitial {
   [_delegate willPresentFullScreenView];
 }
 
-- (void)didShowAd {
-  // Do nothing.
+- (void)interstitialAdDidPresent:(VungleInterstitial *)interstitial {
```

Comment:
Question, do this callback and `interstitialAdDidTrackImpression` call at the same time?

Reply:
No. It should be very close. The DidPresent callback is called when the view is successfully attached, and the impression callback is called when our ad plays successfully and sends a signal to the Vungle SDK to trigger.